Support a dark theme with ?theme=dark

jacob314 opened the pull request:
Generally everything looks good with the dark theme. I need to fix the GPU and CPU label but Kenzie also has a CL modifying that so I'll wait.

The design is to use a ThemedColor class everywhere themed colors are needed from code and use CSS variables for all themed colors in our style sheets combined with an existing dark github theme.
Overall the dark theme looks pretty good although it could use some polish.
Feel free to comment on any colors that seem off or to send out follow up CLs polishing the dark and light themes. Now that the colors we were using are all in one place it should be more obvious where we have diverged unintentionally.

<img width="1564" alt="screen shot 2019-02-11 at 3 51 31 pm" src="https://user-images.githubusercontent.com/1226812/52603491-ec99ae00-2e1b-11e9-9430-251ece78708b.png">
<img width="1571" alt="screen shot 2019-02-11 at 3 51 21 pm" src="https://user-images.githubusercontent.com/1226812/52603492-ec99ae00-2e1b-11e9-897d-6c5358fd3ee0.png">
<img width="1565" alt="screen shot 2019-02-11 at 3 50 55 pm" src="https://user-images.githubusercontent.com/1226812/52603493-ec99ae00-2e1b-11e9-9dbc-d7ddd547114e.png">
<img width="1564" alt="screen shot 2019-02-11 at 3 33 14 pm" src="https://user-images.githubusercontent.com/1226812/52603494-ed324480-2e1b-11e9-902a-3abeddcd2488.png">

jacob314 commented on packages/devtools/web/styles.css:
```diff
@@ -2,16 +2,58 @@
+/* be refactored to eliminate cases where very similar colors are used with */
+/* no clear reason to diverge. */
+/* If you add a CSS variable here, be sure to also add it to styles_dart.css */
+:root {
```
This is the core definition of the color scheme. Some of these entries are a bit redundant. Feel free to suggest cases where slightly different colors should be merged.
Cases where the same color appear twice are mainly cases where the same color should have different versions in the dark theme and vice versa.

kenzieschmoll replied:
super nit: can we alphabetize these?

jacob314 replied:
done. sorted here and in styles_dark.css

jacob314 commented on packages/devtools/web/index.html:
```diff
@@ -24,6 +25,18 @@
     <link rel="stylesheet" href="packages/devtools/src/timeline/timeline.css">
 
+    <script type="text/javascript">
+        (function() {
```
I added this code here to make sure the dark theme is loaded at initial page load
if `theme=dark" is present.
Otherwise you would get a flash of the light theme which would look bad.
I can clean this up to not use document.write
as I should feel bad for using it. :)

kenzieschmoll commented on packages/devtools/lib/src/ui/theme.dart:
```diff
@@ -0,0 +1,72 @@
+import 'flutter_html_shim.dart';
+
+bool _isDarkTheme = false;
+bool get isDarkTheme => _isDarkTheme;
```
nit: reverse order of the getter and the var?

jacob314 replied:
done. Also added a comment.

devoncarew commented:
Some comments just based on the screen-shots:

- for the logging view, we may nee to tweak the flutter.frame length bars, for both selected and non-selected (there's not a lot of contrast between them and the background color)
- in the timeline page, the text for the cpu and gpu labels needs to be darker, to contrast with the label background
- we may want to choose new colors to cycle through for the timeline flamechart bars
- the background of the flamechart seems pretty dark; we may want a slightly lighter color for it
- it would be interesting to see how the zenburn theme looks like for the debugger source area (and if some of those colors could inform other large areas, like the flamechart background)

Those are all nits; I'm happy to play with some color tweaks post-landing.

devoncarew commented:
For an example of the zenburn theme in codemirror, see https://dartpad.dartlang.org/ :)

DanTup commented on packages/devtools/web/styles_dark.css:
```diff
@@ -0,0 +1,48 @@
+    --counter-color: #999;
+    --dark-shadow-color: #999;
+    --default-background-rgb: 30, 30, 30;
+    --default-background: #1e1e1e;
```
Would something like this work to avoid duplicated colours for those we have both RGB/Hex?

```css
    --default-background-rgb: 30, 30, 30;
    --default-background: rgba(var(default-background-rgb), 1);
```

jacob314 replied:
good idea. done.

jacob314 commented on packages/devtools/lib/src/inspector/inspector.css:
```diff
@@ -20,7 +20,7 @@
     overflow: hidden;
     margin-top: 10px;
-    color: #fff;
+    color: var(--default-background);
```
I expect our CSS has a lot of unused junk that can be cleaned up. Removed.

jacob314 commented on packages/devtools/lib/src/inspector/inspector.css:
```diff
@@ -107,7 +107,7 @@ messages beautiful.
 
 .inspector-style-error {
-    background-color: #f97c7c;
+    background-color: var(--error-background);
```
That is a reasonable enhancement. Feel free to add as a followup cl. For now we are fine as by convention any time there isn't a matching var(--foo-foreground) then the regular foreground color is must be ok to use with the background.

jacob314 commented on packages/devtools/lib/src/inspector/inspector_text_styles.dart:
```diff
@@ -1,16 +1,25 @@
-const TextStyle regular = TextStyle(color: Colors.black);
-final TextStyle warning = TextStyle(color: Colors.orange.shade500);
-final TextStyle error = TextStyle(color: Colors.red.shade500);
+final TextStyle grayed = TextStyle(
```
done. changed to faded. This isn't quite disabled based on the current usage.

jacob314 commented on packages/devtools/lib/src/ui/theme.dart:
```diff
@@ -0,0 +1,72 @@
+}
+
+// Commonly used themed colors.
+const ThemedColor white = ThemedColor(Colors.white, Colors.black);
```
Sure. 
Changed to foregroundColor and backgroundColor.

DanTup replied:
Should that change be in this PR, it still seems to be `white`/`black` (I don't think it needs to be done in this PR, just not sure if it got lost).

jacob314 replied:
actually defaultForeground 
and 
defaultBackground.
Happy to change again. This sort of refactor is easy to do.

DanTup replied:
Looks good now - the names you picked seem fine to me, the comment was just that the change didn't seem to be included (I can see it now though) 👍

jacob314 commented on packages/devtools/lib/src/timeline/timeline.css:
```diff
@@ -32,14 +33,16 @@
 }
 
 .timeline-frame.selected .bar {
-    border: 2px solid #A2A2A2;
```
yeah. The wider border bar width when selected looked strange. If it is useful we should add it back but only for the outside border of the combined bar.

jacob314 commented:
I've backed out the github dark theme css file with a restricted.
I've moved the deps to a third_party directory placing each in a stub package we import with a path import so we don't have to publish.

I'd like to land as is and then we can add a dark primer theme when we find or create one.
I've also upgraded to primer 11.0 as it should be easier to create a dark primer theme that matches the latest version. Only one line of our css was broken by upgrading so I fixed it (margins on .gutter which we happened to also use).

@devoncarew those suggestions are all good. Lets deal with them post landing. Even ignoring the issue of finding a compatible primer dark theme, this is really only expected to be a first step towards good dark theme support requiring some small tweaking to make the dark theme be polished.

devoncarew commented on packages/devtools/test/ui/theme_test.dart:
```diff
@@ -0,0 +1,55 @@
+import 'package:test/test.dart';
```
nit: copyright header

jacob314 replied:
done.

devoncarew commented on CONTRIBUTING.md:
```diff
@@ -70,3 +70,11 @@ file output still looks reasonable and execute the following command to update t
+
+### third_party dependencies
+
+All content not authored by the flutter team must go in the third_party
```
s/flutter/Flutter

jacob314 replied:
done
